File: MorseCSDK/src/receiver.c
#include <stdio.h>
#include <string.h>
#include "main.h"

// Message buffer to display on LCD
char display_message[MAX_MESSAGE_LENGTH] = "";
int message_index = 0;
uint8_t actual_lcd_address = LCD_ADDRESS;  // Will be updated during scan
bool lcd_available = false;                // Flag to track if LCD is working

// Morse code patterns for A-Z
const char* morse_code[] = {
    ".-",     // A
    "-...",   // B
    "-.-.",   // C
    "-..",    // D
    ".",      // E
    "..-.",   // F
    "--.",    // G
    "....",   // H
    "..",     // I
    ".---",   // J
    "-.-",    // K
    ".-..",   // L
    "--",     // M
    "-.",     // N
    "---",    // O
    ".--.",   // P
    "--.-",   // Q
    ".-.",    // R
    "...",    // S
    "-",      // T
    "..-",    // U
    "...-",   // V
    ".--",    // W
    "-..-",   // X
    "-.--",   // Y
    "--.."    // Z
};

// Segment encoding for A-Z (7-segment display)
const uint8_t segment_encoding[] = {
    0b1110111,  // A
    0b0011111,  // B
    0b1001110,  // C
    0b0111101,  // D
    0b1001111,  // E
    0b1000111,  // F
    0b1011110,  // G
    0b0110111,  // H
    0b0110000,  // I
    0b0111100,  // J
    0b1110110,  // K
    0b0001110,  // L
    0b1010101,  // M
    0b0010101,  // N
    0b1111110,  // O
    0b1100111,  // P
    0b1110011,  // Q
    0b0000101,  // R
    0b1011011,  // S
    0b0001111,  // T
    0b0111110,  // U
    0b0111110,  // V
    0b1111010,  // W
    0b0110111,  // X
    0b0111011,  // Y
    0b1101101   // Z
};

// Forward declarations
void uart_log(const char* message);

// I2C helper functions
bool i2c_scan_for_lcd() {
    printf("Scanning I2C bus for devices...\n");
    printf("Using SDA_PIN: %d, SCL_PIN: %d\n", SDA_PIN, SCL_PIN);
    
    bool device_found = false;
    
    for (uint8_t addr = 0; addr < 128; addr++) {
        // Skip reserved addresses
        if (addr >= 0x00 && addr <= 0x07) continue;
        if (addr >= 0x78 && addr <= 0x7F) continue;
        
        uint8_t dummy = 0;
        int result = i2c_write_blocking(i2c0, addr, &dummy, 1, false);
        
        if (result >= 0) {
            printf("I2C device found at address 0x%02X\n", addr);
            device_found = true;
            
            // If we find our expected LCD address, use it
            if (addr == LCD_ADDRESS) {
                printf("Found LCD at expected address 0x%02X\n", addr);
                actual_lcd_address = addr;
                return true;
            } else if (actual_lcd_address == LCD_ADDRESS) {
                // Otherwise use the first device we find
                printf("Assuming device at 0x%02X is the LCD\n", addr);
                actual_lcd_address = addr;
            }
        }
    }
    
    if (!device_found) {
        printf("No I2C devices found! Check connections.\n");
        return false;
    }
    
    printf("Using I2C address 0x%02X for LCD\n", actual_lcd_address);
    return true;
}

// Basic I2C write
bool lcd_write_byte(uint8_t byte_value) {
    int result = i2c_write_blocking(i2c0, actual_lcd_address, &byte_value, 1, false);
    return result == 1;
}

// LCD functions
void lcd_pulse_enable(uint8_t data) {
    // Send with enable bit high
    lcd_write_byte(data | LCD_EN_BIT | LCD_BACKLIGHT);
    sleep_us(500);  // Enable pulse must be >450ns
    
    // Send with enable bit low
    lcd_write_byte((data & ~LCD_EN_BIT) | LCD_BACKLIGHT);
    sleep_us(100);  // Commands need time to settle
}

// Write 4 bits to LCD
void lcd_write_4bits(uint8_t value) {
    uint8_t data = (value & 0xF0) | LCD_BACKLIGHT;  // High nibble, backlight on
    lcd_pulse_enable(data);
}

// Write a command to the LCD
void lcd_command(uint8_t command) {
    // High nibble with RS=0 (command mode)
    uint8_t high_nibble = (command & 0xF0) | LCD_BACKLIGHT;
    lcd_pulse_enable(high_nibble);
    
    // Low nibble with RS=0 (command mode)
    uint8_t low_nibble = ((command << 4) & 0xF0) | LCD_BACKLIGHT;
    lcd_pulse_enable(low_nibble);
    
    // Most commands need >37us to settle
    sleep_ms(2);
}

// Write data to the LCD
void lcd_data(uint8_t data) {
    // High nibble with RS=1 (data mode)
    uint8_t high_nibble = (data & 0xF0) | LCD_RS_BIT | LCD_BACKLIGHT;
    lcd_pulse_enable(high_nibble);
    
    // Low nibble with RS=1 (data mode)
    uint8_t low_nibble = ((data << 4) & 0xF0) | LCD_RS_BIT | LCD_BACKLIGHT;
    lcd_pulse_enable(low_nibble);
    
    sleep_us(100);
}

bool lcd_init() {
    printf("Initializing I2C for LCD...\n");
    
    // Initialize I2C at a slower speed for reliability
    i2c_init(i2c0, 100 * 1000);  // 100 kHz
    
    // Set up I2C pins
    gpio_set_function(SDA_PIN, GPIO_FUNC_I2C);
    gpio_set_function(SCL_PIN, GPIO_FUNC_I2C);
    gpio_pull_up(SDA_PIN);
    gpio_pull_up(SCL_PIN);
    
    // Scan for I2C devices
    if (!i2c_scan_for_lcd()) {
        printf("Failed to find LCD on I2C bus\n");
        return false;
    }
    
    // Wait for LCD to power up completely
    sleep_ms(50);
    
    // Turn on backlight
    lcd_write_byte(LCD_BACKLIGHT);
    sleep_ms(50);
    
    // According to datasheet, we need to send 8-bit mode command 3 times
    // Send 0011
    lcd_write_4bits(0x30);
    sleep_ms(5);
    
    // Send 0011 again
    lcd_write_4bits(0x30);
    sleep_ms(5);
    
    // Send 0011 third time
    lcd_write_4bits(0x30);
    sleep_ms(5);
    
    // Now switch to 4-bit mode
    lcd_write_4bits(0x20);
    sleep_ms(5);
    
    // Function set: 4-bit mode, 2 lines, 5x8 dots
    lcd_command(LCD_FUNCTIONSET | LCD_4BITMODE | LCD_2LINE | LCD_5x8DOTS);
    
    // Display control: display on, cursor off, blink off
    lcd_command(LCD_DISPLAYCONTROL | LCD_DISPLAYON | LCD_CURSOROFF | LCD_BLINKOFF);
    
    // Clear display
    lcd_command(LCD_CLEARDISPLAY);
    sleep_ms(5);  // Clear command needs extra time (>1.52ms)
    
    // Entry mode: increment cursor position, no display shift
    lcd_command(LCD_ENTRYMODESET | LCD_ENTRYLEFT | LCD_ENTRYSHIFTDECREMENT);
    
    // Return home
    lcd_command(LCD_RETURNHOME);
    sleep_ms(5);  // Return home needs extra time
    
    // Test with a simple message
    const char* test_msg = "LCD Ready";
    for (int i = 0; test_msg[i] != '\0'; i++) {
        lcd_data(test_msg[i]);
    }
    
    lcd_available = true;
    printf("LCD initialization complete\n");
    return true;
}

void lcd_clear() {
    if (!lcd_available) return;
    lcd_command(LCD_CLEARDISPLAY);
    sleep_ms(5);  // Clear needs extra time
}

void lcd_set_cursor(uint8_t col, uint8_t row) {
    if (!lcd_available) return;
    
    static const uint8_t row_offsets[] = {0x00, 0x40, 0x14, 0x54};
    lcd_command(LCD_SETDDRAMADDR | (col + row_offsets[row % 4]));
}

void lcd_print(const char* text) {
    if (!lcd_available) return;
    
    while (*text) {
        lcd_data(*text++);
    }
}

void add_to_message(char letter) {
    if (message_index >= MAX_MESSAGE_LENGTH - 1) {
        // Shift message to make room for new character
        memmove(display_message, display_message + 1, MAX_MESSAGE_LENGTH - 2);
        message_index = MAX_MESSAGE_LENGTH - 2;
    }
    
    display_message[message_index++] = letter;
    display_message[message_index] = '\0';
    
    update_lcd_display();
}

void update_lcd_display() {
    if (!lcd_available) return;
    
    // Calculate how much of the message to show (last 16 chars)
    int start_pos = 0;
    if (strlen(display_message) > LCD_CHAR_WIDTH) {
        start_pos = strlen(display_message) - LCD_CHAR_WIDTH;
    }
    
    // Display title and message
    lcd_clear();
    lcd_set_cursor(0, 0);
    lcd_print("Morse Receiver");
    
    lcd_set_cursor(0, 1);
    lcd_print(display_message + start_pos);
    
    // Log what we're displaying
    char debug_msg[64];
    sprintf(debug_msg, "LCD Display: %s", display_message);
    uart_log(debug_msg);
}

void init_receiver() {
    stdio_init_all();
    sleep_ms(2000);  // Give time for USB to initialize
    
    printf("\n\n==== Morse Code Receiver with LCD ====\n");
    
    // Initialize UART
    uart_init(UART_ID, BAUD_RATE);
    gpio_set_function(UART_RX_PIN, GPIO_FUNC_UART);
    gpio_set_function(UART_TX_PIN, GPIO_FUNC_UART);
    
    // Initialize onboard LED
    gpio_init(PICO_DEFAULT_LED_PIN);
    gpio_set_dir(PICO_DEFAULT_LED_PIN, GPIO_OUT);
    
    // Initialize the LCD
    if (lcd_init()) {
        printf("LCD successfully initialized\n");
    } else {
        printf("LCD initialization failed - will continue without LCD\n");
    }
    
    printf("Receiver initialized and ready\n");
}

void uart_log(const char* message) {
    printf("%s\n", message);
}

char decode_morse(const char* morse) {
    if (!morse || strlen(morse) == 0) return '?';
    
    for (int i = 0; i < 26; i++) {
        if (strcmp(morse_code[i], morse) == 0) {
            return 'A' + i;
        }
    }
    return '?';
}

void display_letter(char letter) {
    if (letter == ' ') {
        uart_log("Detected: SPACE");
        add_to_message(' ');
        return;
    }
    
    int index = letter - 'A';
    if (index >= 0 && index < 26) {
        char msg[32];
        sprintf(msg, "Decoded: %c", letter);
        uart_log(msg);
        
        // Display on LCD
        add_to_message(letter);
    } else {
        uart_log("Invalid letter detected");
    }
}

int main() {
    init_receiver();
    
    char morse_buffer[MAX_MORSE_LENGTH] = {0};
    int buffer_index = 0;
    uint32_t last_signal_time = 0;
    uint32_t last_event_time = 0;
    bool in_character = false;
    bool in_word = false;
    bool space_added = false;
    
    uart_log("Starting Morse reception...");
    
    // Display welcome message on LCD
    if (lcd_available) {
        lcd_clear();
        lcd_set_cursor(0, 0);
        lcd_print("Morse Receiver");
        lcd_set_cursor(0, 1);
        lcd_print("Waiting...");
    }
    
    while (true) {
        uint32_t current_time = to_ms_since_boot(get_absolute_time());
        
        // Check for incoming data
        if (uart_is_readable(UART_ID)) {
            char c = uart_getc(UART_ID);
            
            // Flash LED when receiving data
            gpio_put(PICO_DEFAULT_LED_PIN, 1);
            sleep_ms(5);
            gpio_put(PICO_DEFAULT_LED_PIN, 0);
            
            if (c == '.' || c == '-') {
                // Start a new character if we're not in one
                if (!in_character) {
                    buffer_index = 0;
                    memset(morse_buffer, 0, sizeof(morse_buffer));
                    in_character = true;
                    space_added = false;
                }
                
                // Add the dot or dash to our buffer
                if (buffer_index < MAX_MORSE_LENGTH - 1) {
                    morse_buffer[buffer_index++] = c;
                    morse_buffer[buffer_index] = '\0';
                    
                    char msg[32];
                    sprintf(msg, "Received signal: %c", c);
                    uart_log(msg);
                }
                
                last_signal_time = current_time;
                last_event_time = current_time;
                in_word = true;
            }
            // Character gap marker directly from transmitter
            else if (strncmp(&c, "C", 1) == 0 || strncmp(&c, "c", 1) == 0) {
                if (strlen(morse_buffer) > 0 && in_character) {
                    // We have a complete character to decode
                    char decoded = decode_morse(morse_buffer);
                    if (decoded != '?') {
                        display_letter(decoded);
                        char msg[64];
                        sprintf(msg, "Decoded character: %c (%s)", decoded, morse_buffer);
                        uart_log(msg);
                    } else {
                        char msg[64];
                        sprintf(msg, "Failed to decode: (%s)", morse_buffer);
                        uart_log(msg);
                    }
                    
                    // Reset the buffer
                    memset(morse_buffer, 0, MAX_MORSE_LENGTH);
                    buffer_index = 0;
                    in_character = false;
                }
                last_event_time = current_time;
            }
            // Word gap marker directly from transmitter
            else if (strncmp(&c, "W", 1) == 0 || strncmp(&c, "w", 1) == 0) {
                if (in_word && !space_added) {
                    display_letter(' ');
                    in_word = false;
                    space_added = true;
                    uart_log("Word gap detected - adding space");
                }
                last_event_time = current_time;
            }
            // Handle entire CHAR GAP or WORD GAP strings
            else if (c == 'H' && in_character) { // Part of "CHAR GAP"
                // Finish current character if we have content
                if (strlen(morse_buffer) > 0) {
                    char decoded = decode_morse(morse_buffer);
                    if (decoded != '?') {
                        display_letter(decoded);
                    }
                    memset(morse_buffer, 0, MAX_MORSE_LENGTH);
                    buffer_index = 0;
                    in_character = false;
                }
            }
            else if (c == 'O' && in_word) { // Part of "WORD GAP"
                if (!space_added) {
                    display_letter(' ');
                    space_added = true;
                }
                in_word = false;
            }
        }
        
        // Automatic timeout detection for incomplete characters
        if (in_character && (current_time - last_signal_time > INTER_CHAR_GAP)) {
            if (strlen(morse_buffer) > 0) {
                char decoded = decode_morse(morse_buffer);
                if (decoded != '?') {
                    display_letter(decoded);
                    char msg[64];
                    sprintf(msg, "Auto-decoded by timeout: %c (%s)", decoded, morse_buffer);
                    uart_log(msg);
                } else {
                    char msg[64];
                    sprintf(msg, "Failed to auto-decode: (%s)", morse_buffer);
                    uart_log(msg);
                }
                
                memset(morse_buffer, 0, MAX_MORSE_LENGTH);
                buffer_index = 0;
            }
            in_character = false;
        }
        
        // Detect word gaps automatically if no events for a while
        if (in_word && !space_added && (current_time - last_event_time > WORD_GAP)) {
            display_letter(' ');
            space_added = true;
            in_word = false;
            uart_log("Auto word gap - adding space");
        }
        
        sleep_ms(5);  // Prevent tight polling
    }
    
    return 0;
}

File: MorseRSDK/src/bin/receiver.rs
#![no_std]
#![no_main]

#[link_section = ".boot2"]
#[used]
pub static BOOT2: [u8; 256] = rp2040_boot2::BOOT_LOADER_GENERIC_03H;

use panic_halt as _;
use core::fmt::Write;
use core::result::Result::{Ok, Err};
use core::iter::Iterator;
use core::convert::Into;
use core::write;

use rp2040_hal::{
    gpio::{bank0::Gpio25, Pin, FunctionSio, SioOutput, PullDown, FunctionI2c},
    pac,
    timer::Timer,
    clocks::{Clock, init_clocks_and_plls},
    watchdog::Watchdog,
    Sio,
    gpio::Pins,
    uart::{UartPeripheral, UartConfig, DataBits, Parity, StopBits},
    i2c::I2C,
};
use rp_pico::XOSC_CRYSTAL_FREQ;
use rp2040_hal::entry;
use cortex_m::delay::Delay;
use heapless::String;
use nb::block;
use embedded_hal::digital::OutputPin;
use embedded_hal_0_2::serial;
use embedded_hal_0_2::blocking::i2c;
use rp2040_hal::fugit::RateExtU32;

use morse_rsdk::{
    BAUD_RATE,
    MORSE_CODE, LCD_ADDRESS, LCD_BACKLIGHT, LCD_EN_BIT, LCD_RS_BIT, LCD_CLEARDISPLAY,
    LCD_RETURNHOME, LCD_ENTRYMODESET, LCD_DISPLAYCONTROL, LCD_FUNCTIONSET,
    LCD_SETDDRAMADDR, LCD_ENTRYLEFT, LCD_ENTRYSHIFTDECREMENT,
    LCD_DISPLAYON, LCD_CURSOROFF, LCD_BLINKOFF, LCD_4BITMODE, LCD_2LINE, LCD_5X8_DOTS,
    LCD_CHAR_WIDTH, MAX_MESSAGE_LENGTH, MAX_MORSE_LENGTH
};

// Timing constants
const INTER_CHAR_GAP: u32 = 1000; // 1 second for inter-character gap
const WORD_GAP: u32 = 2000; // 2 seconds for word gap

pub struct Receiver<UART, I2C> 
where
    UART: serial::Write<u8> + serial::Read<u8>,
    I2C: i2c::Write,
{
    uart: UART,
    i2c: I2C,
    led_pin: Pin<Gpio25, FunctionSio<SioOutput>, PullDown>,
    timer: Timer,
    delay: Delay,
    display_message: String<{MAX_MESSAGE_LENGTH}>,
    message_index: usize,
    actual_lcd_address: u8,
    lcd_available: bool,
}

impl<UART, I2C> Receiver<UART, I2C>
where
    UART: serial::Write<u8> + serial::Read<u8>,
    I2C: i2c::Write,
{
    pub fn new(
        uart: UART,
        i2c: I2C,
        led_pin: Pin<Gpio25, FunctionSio<SioOutput>, PullDown>,
        timer: Timer,
        delay: Delay,
    ) -> Self {
        Self {
            uart,
            i2c,
            led_pin,
            timer,
            delay,
            display_message: String::new(),
            message_index: 0,
            actual_lcd_address: LCD_ADDRESS,
            lcd_available: false,
        }
    }

    pub fn init(&mut self) {
        // Test LED first
        self.led_pin.set_high().unwrap();
        self.delay.delay_ms(500);
        self.led_pin.set_low().unwrap();
        
        // Test UART
        self.uart_log("UART test - if you see this, UART works");
        
        // Test I2C LCD
        if self.lcd_init() {
            self.uart_log("LCD initialized successfully");
            self.lcd_command(LCD_CLEARDISPLAY);
            self.lcd_print("Hello World!");
        } else {
            self.uart_log("LCD init failed");
        }
        
        self.uart_log("System ready");
    }

    fn lcd_write_byte(&mut self, byte_value: u8) -> bool {
        self.i2c.write(self.actual_lcd_address, &[byte_value]).is_ok()
    }

    fn lcd_pulse_enable(&mut self, data: u8) {
        self.lcd_write_byte(data | LCD_EN_BIT | LCD_BACKLIGHT);
        self.delay.delay_us(500);
        self.lcd_write_byte((data & !LCD_EN_BIT) | LCD_BACKLIGHT);
        self.delay.delay_us(100);
    }

    fn lcd_write_4bits(&mut self, value: u8) {
        let data = (value & 0xF0) | LCD_BACKLIGHT;
        self.lcd_pulse_enable(data);
    }

    fn lcd_command(&mut self, command: u8) {
        let high_nibble = (command & 0xF0) | LCD_BACKLIGHT;
        self.lcd_pulse_enable(high_nibble);
        
        let low_nibble = ((command << 4) & 0xF0) | LCD_BACKLIGHT;
        self.lcd_pulse_enable(low_nibble);
        
        self.delay.delay_ms(2);
    }

    fn lcd_data(&mut self, data: u8) {
        let high_nibble = (data & 0xF0) | LCD_RS_BIT | LCD_BACKLIGHT;
        self.lcd_pulse_enable(high_nibble);
        
        let low_nibble = ((data << 4) & 0xF0) | LCD_RS_BIT | LCD_BACKLIGHT;
        self.lcd_pulse_enable(low_nibble);
        
        self.delay.delay_us(100);
    }

    fn lcd_init(&mut self) -> bool {
        self.uart_log("Initializing I2C for LCD...");
        
        self.delay.delay_ms(50);
        self.lcd_write_byte(LCD_BACKLIGHT);
        self.delay.delay_ms(50);
        
        self.lcd_write_4bits(0x30);
        self.delay.delay_ms(5);
        self.lcd_write_4bits(0x30);
        self.delay.delay_ms(5);
        self.lcd_write_4bits(0x30);
        self.delay.delay_ms(5);
        
        self.lcd_write_4bits(0x20);
        self.delay.delay_ms(5);
        
        self.lcd_command(LCD_FUNCTIONSET | LCD_4BITMODE | LCD_2LINE | LCD_5X8_DOTS);
        self.lcd_command(LCD_DISPLAYCONTROL | LCD_DISPLAYON | LCD_CURSOROFF | LCD_BLINKOFF);
        self.lcd_command(LCD_CLEARDISPLAY);
        self.delay.delay_ms(5);
        self.lcd_command(LCD_ENTRYMODESET | LCD_ENTRYLEFT | LCD_ENTRYSHIFTDECREMENT);
        self.lcd_command(LCD_RETURNHOME);
        self.delay.delay_ms(5);
        
        let test_msg = "LCD Ready";
        for c in test_msg.chars() {
            self.lcd_data(c as u8);
        }
        
        self.lcd_available = true;
        self.uart_log("LCD initialization complete");
        true
    }

    fn lcd_clear(&mut self) {
        if !self.lcd_available { return; }
        self.lcd_command(LCD_CLEARDISPLAY);
        self.delay.delay_ms(5);
    }

    fn lcd_set_cursor(&mut self, col: u8, row: u8) {
        if !self.lcd_available { return; }
        
        let row_offsets = [0x00, 0x40, 0x14, 0x54];
        self.lcd_command(LCD_SETDDRAMADDR | (col + row_offsets[row as usize % 4]));
    }

    fn lcd_print(&mut self, text: &str) {
        if !self.lcd_available { return; }
        
        for c in text.chars() {
            self.lcd_data(c as u8);
        }
    }

    fn add_to_message(&mut self, letter: char) {
        if self.message_index >= MAX_MESSAGE_LENGTH - 1 {
            let mut temp = String::<{MAX_MESSAGE_LENGTH}>::new();
            let _ = write!(&mut temp, "{}", &self.display_message[1..]);
            self.display_message.clear();
            let _ = self.display_message.push_str(temp.as_str());
            self.message_index = MAX_MESSAGE_LENGTH - 2;
        }
        
        let _ = self.display_message.push(letter);
        self.message_index += 1;
        self.update_lcd_display();
    }

    fn update_lcd_display(&mut self) {
        if !self.lcd_available { return; }
        
        let start_pos = if self.display_message.len() > LCD_CHAR_WIDTH {
            self.display_message.len() - LCD_CHAR_WIDTH
        } else {
            0
        };
        
        let display_text = {
            let mut temp = String::<{LCD_CHAR_WIDTH}>::new();
            let _ = write!(&mut temp, "{}", &self.display_message[start_pos..]);
            temp
        };
        
        self.lcd_clear();
        self.lcd_set_cursor(0, 0);
        self.lcd_print("Morse Receiver");
        
        self.lcd_set_cursor(0, 1);
        self.lcd_print(display_text.as_str());
        
        let mut log_msg = String::<128>::new();
        let _ = write!(log_msg, "LCD Display: {}", display_text);
        self.uart_log(log_msg.as_str());
    }

    pub fn uart_log(&mut self, message: &str) {
        for byte in message.as_bytes() {
            let _ = block!(self.uart.write(*byte));
        }
        let _ = block!(self.uart.write(b'\r'));
        let _ = block!(self.uart.write(b'\n'));
    }

    fn decode_morse(&self, morse: &str) -> char {
        if morse.is_empty() { return '?'; }
        
        for (i, code) in MORSE_CODE.iter().enumerate() {
            if *code == morse {
                return (b'A' + i as u8) as char;
            }
        }
        '?'
    }

    fn display_letter(&mut self, letter: char) {
        if letter == ' ' {
            self.uart_log("Detected: SPACE");
            self.add_to_message(' ');
            return;
        }
        
        if letter.is_ascii_uppercase() {
            let mut message = String::<32>::new();
            let _ = write!(&mut message, "Decoded: {}", letter);
            self.uart_log(message.as_str());
            self.add_to_message(letter);
        } else {
            self.uart_log("Invalid letter detected");
        }
    }

    pub fn run(&mut self) {
        let mut morse_buffer = [0u8; MAX_MORSE_LENGTH];
        let mut buffer_index = 0;
        let mut last_signal_time = 0u64;
        let mut last_event_time = 0u64;
        let mut in_character = false;
        let mut in_word = false;
        let mut space_added = false;
        
        self.uart_log("Starting Morse reception...");
        
        if self.lcd_available {
            self.lcd_clear();
            self.lcd_set_cursor(0, 0);
            self.lcd_print("Morse Receiver");
            self.lcd_set_cursor(0, 1);
            self.lcd_print("Waiting...");
        }
        
        loop {
            let current_time = self.timer.get_counter().ticks();
            
            if let Ok(c) = block!(self.uart.read()) {
                self.led_pin.set_high().unwrap();
                self.delay.delay_ms(5);
                self.led_pin.set_low().unwrap();
                
                if c == b'.' || c == b'-' {
                    if !in_character {
                        buffer_index = 0;
                        morse_buffer = [0u8; MAX_MORSE_LENGTH];
                        in_character = true;
                        space_added = false;
                    }
                    
                    if buffer_index < MAX_MORSE_LENGTH - 1 {
                        morse_buffer[buffer_index] = c;
                        buffer_index += 1;
                        
                        let mut message = String::<32>::new();
                        let _ = write!(&mut message, "Received signal: {}", c as char);
                        self.uart_log(message.as_str());
                    }
                    
                    last_signal_time = current_time;
                    last_event_time = current_time;
                    in_word = true;
                }
                else if c == b'C' || c == b'c' {
                    if buffer_index > 0 && in_character {
                        let morse_str = core::str::from_utf8(&morse_buffer[..buffer_index]).unwrap_or("");
                        let decoded = self.decode_morse(morse_str);
                        if decoded != '?' {
                            self.display_letter(decoded);
                            let mut message = String::<64>::new();
                            let _ = write!(&mut message, "Decoded character: {} ({})", decoded, morse_str);
                            self.uart_log(message.as_str());
                        } else {
                            let mut message = String::<64>::new();
                            let _ = write!(&mut message, "Failed to decode: ({})", morse_str);
                            self.uart_log(message.as_str());
                        }
                        
                        morse_buffer = [0u8; MAX_MORSE_LENGTH];
                        buffer_index = 0;
                        in_character = false;
                    }
                    last_event_time = current_time;
                }
                else if c == b'W' || c == b'w' {
                    if in_word && !space_added {
                        self.display_letter(' ');
                        in_word = false;
                        space_added = true;
                        self.uart_log("Word gap detected - adding space");
                    }
                    last_event_time = current_time;
                }
                else if c == b'H' && in_character {
                    if buffer_index > 0 {
                        let morse_str = core::str::from_utf8(&morse_buffer[..buffer_index]).unwrap_or("");
                        let decoded = self.decode_morse(morse_str);
                        if decoded != '?' {
                            self.display_letter(decoded);
                        }
                        morse_buffer = [0u8; MAX_MORSE_LENGTH];
                        buffer_index = 0;
                        in_character = false;
                    }
                }
                else if c == b'O' && in_word {
                    if !space_added {
                        self.display_letter(' ');
                        space_added = true;
                    }
                    in_word = false;
                }
            }
            
            if in_character && (current_time - last_signal_time > INTER_CHAR_GAP.into()) {
                if buffer_index > 0 {
                    let morse_str = core::str::from_utf8(&morse_buffer[..buffer_index]).unwrap_or("");
                    let decoded = self.decode_morse(morse_str);
                    if decoded != '?' {
                        self.display_letter(decoded);
                        let mut message = String::<64>::new();
                        let _ = write!(&mut message, "Auto-decoded by timeout: {} ({})", decoded, morse_str);
                        self.uart_log(message.as_str());
                    } else {
                        let mut message = String::<64>::new();
                        let _ = write!(&mut message, "Failed to auto-decode: ({})", morse_str);
                        self.uart_log(message.as_str());
                    }
                    
                    morse_buffer = [0u8; MAX_MORSE_LENGTH];
                    buffer_index = 0;
                }
                in_character = false;
            }
            
            if in_word && !space_added && (current_time - last_event_time > WORD_GAP.into()) {
                self.display_letter(' ');
                space_added = true;
                in_word = false;
                self.uart_log("Auto word gap - adding space");
            }
            
            self.delay.delay_ms(5);
        }
    }
}

#[entry]
fn main() -> ! {
    let mut pac = pac::Peripherals::take().unwrap();
    let core = pac::CorePeripherals::take().unwrap();
    
    let mut watchdog = Watchdog::new(pac.WATCHDOG);
    let clocks = init_clocks_and_plls(
        XOSC_CRYSTAL_FREQ,
        pac.XOSC,
        pac.CLOCKS,
        pac.PLL_SYS,
        pac.PLL_USB,
        &mut pac.RESETS,
        &mut watchdog,
    ).ok().unwrap();
    
    let sio = Sio::new(pac.SIO);
    let pins = Pins::new(
        pac.IO_BANK0,
        pac.PADS_BANK0,
        sio.gpio_bank0,
        &mut pac.RESETS,
    );
    
    let delay = cortex_m::delay::Delay::new(core.SYST, clocks.system_clock.freq().to_Hz());
    let timer = Timer::new(pac.TIMER, &mut pac.RESETS, &clocks);
    
    let uart = UartPeripheral::new(
        pac.UART0,
        (pins.gpio0.into_function(), pins.gpio1.into_function()),
        &mut pac.RESETS,
    )
    .enable(
        UartConfig::new(BAUD_RATE.Hz(), DataBits::Eight, core::prelude::v1::Some(Parity::Odd), StopBits::One),
        clocks.peripheral_clock.freq(),
    )
    .unwrap();
    
    let i2c = I2C::i2c0(
        pac.I2C0,
        pins.gpio4.into_pull_up_input().into_function::<FunctionI2c>(),
        pins.gpio5.into_pull_up_input().into_function::<FunctionI2c>(),
        100_000.Hz(),
        &mut pac.RESETS,
        clocks.system_clock.freq(),
    );
    
    let led_pin = pins.gpio25.into_push_pull_output();
    
    let mut receiver = Receiver::new(
        uart,
        i2c,
        led_pin,
        timer,
        delay,
    );
    
    receiver.init();
    receiver.run();
    
    loop {}
}